### ISSUE TITLE
Optional dependency with a native module is not installed

### ISSUE THREAD

**Issue by nebrius** [yarnpkg/yarn#670]
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

When I install [johnny-five](https://github.com/rwaldron/johnny-five), it's [optional dependency](https://github.com/rwaldron/johnny-five/blob/master/package.json#L139) [serialport](https://github.com/EmergingTechnologyAdvisors/node-serialport) with a native module (using node-pre-gyp) is missing from `node_modules`, but installing serialport directly works fine.

**If the current behavior is a bug, please provide the steps to reproduce.**

Using this package.json file:

```json
{
  "name": "yarntest",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "johnny-five": "^0.10.0"
  },
  "author": "",
  "license": "ISC"
}
```

1. `yarn cache clean`
1. `rm yarn.lock`
1. `yarn install`
1. `cd node_modules`
1. `find . -name serialport` returns nothing

But, using this package.json file:

```json
{
  "name": "yarntest",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "serialport": "^4.0.0"
  },
  "author": "",
  "license": "ISC"
}
```

1. `yarn cache clean`
1. `rm yarn.lock`
1. `yarn install`
1. `cd node_modules`
1. `find . -name serialport` shows it has been installed

**What is the expected behavior?**

Serialport should be installed as a dependency (the npm client works this way).

**Please mention your node.js, yarn and operating system version.**

Node: 6.7.0
Yarn: 0.15.1
OS: Ubuntu 14.04 x64

**Comment by ljharb** [comment]
To clarify, optionalDeps are just like deps except that a failure to install or compile won't kill the overall install.

**Comment by silverwind** [comment]
Seeing the same, it looks like optional dependencies are not attempted to be built. A manual `yarn add <pkg>` does compile it fine tho.

**Comment by kittens** [comment]
Fixed via #789.

**Comment by naholyr** [comment]
confirmed here:

* `yarn add hiredis --optional` (properly adds to package.json under optionalDependencies)
* `rm -rf node_modules`
* `yarn`
* `node -e 'require("hiredis")'` (fails: module not found)

**Comment by shepmaster** [comment]
> optionalDeps are just like deps except that a failure to install or compile won't kill the overall install.

Is there some way of knowing if a failure to compile has occurred? For example, my output looks like:

```
$ yarn
yarn install v0.15.1
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 📃  Building fresh packages...
success Saved lockfile.
✨  Done in 1.28s.
```

No obvious errors are shown.

**Comment by shepmaster** [comment]
This also applies to optional dependencies in our own `package.json`:

```js
{
  "name": "scratch",
  "version": "0.0.1",
  "description": "Reproduction",
  "main": "index.js",
  "license": "MIT",
  "optionalDependencies": {
    "node-zopfli": "^2.0.1"
  }
}
```

```
$ rm -rf node_modules
$ yarn
yarn install v0.15.1
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 📃  Building fresh packages...
success Saved lockfile.
✨  Done in 1.34s.
$ find node_modules -name '*zopfli*'
node_modules/.bin/zopfli
node_modules/.bin/zopflipng
```

It is interesting to note that the *binaries* get installed, but the library itself doesn't. Running the binaries fails due to unsatisfied dependencies.

**Comment by shepmaster** [comment]
Checking out the package.json for node-zopfli, there's an `install` script:

```js
{
  "scripts": {
    "install": "node-pre-gyp install --fallback-to-build",
  },
}
```

Which... has a space in it! I think that at the root, this issue is yet another thing fixed by #809. The unique problem here might be the failures to run `install` is silently ignored because it is an optional dependency.

Following the instructions in [this comment](https://github.com/yarnpkg/yarn/pull/809#issuecomment-253346655) and [this comment](https://github.com/yarnpkg/yarn/pull/809#issuecomment-253413690) to run the master branch of yarn:

```
git clone git://github.com/yarnpkg/yarn.git
cd yarn
yarn install
npm run build
npm link
```

I see that the native dependency is now installed.

Now we just gotta wait for that commit to be released in an official version.

**Comment by smithrobs** [comment]
Pulling master of yarn as @shepmaster suggested resolved this for me as well (my culprit was noble and xpc-connection). 👍

**Comment by ashconnell** [comment]
Which commit was this fixed in?

**Comment by shepmaster** [comment]
> Which commit was this fixed in?

Please [read the previous comment](https://github.com/yarnpkg/yarn/issues/670#issuecomment-253818631):

> I think that at the root, this issue is yet another thing fixed by #809.

**Comment by michel-kraemer** [comment]
This issue seems to be fixed in 0.16.0.

**Comment by nebrius** [comment]
Confirmed this is fixed in 0.16.1, thanks!
